### DISCUSSION TITLE
Creating a class for HVAC equipment

### DISCUSSION THREAD

**Discussion by michpro** [grodansparadis/vscp#277] [Ideas]
What do you think about creating a separate protocol class for events generated by HVAC equipment, based on OpenTherm, KM-BUS (Viessmann) or other specifications? Of course, this can already be done, using events from the classes: control, information, measurement,... but if this was collected in one class I think it would be more unambiguous.

**Comment by gedass2000** [comment]
I had idea to develop similar CAN  VSCP adapter but my Viessmann Vitodens 100 gas boiler has very limited version of the OpenTherm protocol.

**Comment by grodansparadis** [comment]
I think that can be a good idea.  For VSCP  if one want to use **class.control,type. on** to turn something **off** that is OK.  It's up to the implementor to decide how he/she use the events even if such a choice may be confusing for most of us .  The same is true for  events in other classes. Duplicates are OK and there will always be many different  ways to use the events that are available, and even construct your own. "You have to know what to expect to interpret what you get". Thats the only rule.

So it is OK and a good idea to construct a class for HVAC equipment as long as it is general and not vendor specific and is needed.  Please suggest how you think it should be built and the types that would be needed. Not my field.  A "HVAC ON" and "HVAC OFF" without some extra special data may IMHO don't contribute much to the stuff that is already available.  But you have probably other thoughts an ideas on the subject I guess. Let them flow. :)

The only area I am not so keen to have variants in is for measurements. It destroy a lot of the plug and play thought behind the construct.  You will have to argue hard to add a type that is just used for a measurement and is not defined in the measurement classes.

**Comment by michpro** [comment]
These are events, defined in the OpenTherm protocol (it seems to have a range of events covering all possible situations in HVAC systems)
| LP | Event | Data type [2B] | Note |
| :-- | :--- | :-----: | :-------|
| 1 | Status | flag8/flag8 | Master and Slave Status flags.
| 2 | TSet | f8.8 | Control Setpoint i.e.CH water temperature Setpoint(°C)
| 3 | MConfigMMemberIDcode | flag8/u8 | Master Configuration Flags / Master MemberID Code
| 4 | SConfigSMemberIDcode | flag8/u8 | Slave Configuration Flags / Slave MemberID Code
| 5 | RemoteRequest | u8/u8 | Remote Request
| 6 | ASFflags | flag8/u8 | Application - specific fault flags and OEM fault code
| 7 | RBPflags | flag8/flag8 | Remote boiler parameter transfer - enable & read / write flags
| 8 | CoolingControl | f8.8 | Cooling control signal(%)
| 9 | TsetCH2 | f8.8 | Control Setpoint for 2e CH circuit(°C)
| 10 | TrOverride | f8.8 | Remote override room Setpoint
| 11 | TSP | u8/u8 | Number of Transparent - Slave - Parameters supported by slave
| 12 | TSPindexTSPvalue | u8/u8 | Index number / Value of referred - to transparent slave parameter.
| 13 | FHBsize | u8/u8 | Size of Fault - History - Buffer supported by slave
| 14 | FHBindexFHBvalue | u8/u8 | Index number / Value of referred - to fault - history buffer entry.
| 15 | MaxRelModLevelSetting | f8.8 | Maximum relative modulation level setting(%)
| 16 | MaxCapacityMinModLevel | u8/u8 | Maximum boiler capacity(kW) / Minimum boiler modulation level(%)
| 17 | TrSet | f8.8 | Room Setpoint(°C)
| 18 | RelModLevel | f8.8 | Relative Modulation Level(%)
| 19 | CHPressure | f8.8 | Water pressure in CH circuit(bar)
| 20 | DHWFlowRate | f8.8 | Water flow rate in DHW circuit. (litres / minute)
| 21 | DayTime | special/u8 | Day of Week and Time of Day
| 22 | Date | u8/u8 | Calendar date
| 23 | Year | u16 | Calendar year
| 24 | TrSetCH2 | f8.8 | Room Setpoint for 2nd CH circuit(°C)
| 25 | Tr | f8.8 | Room temperature(°C)
| 26 | Tboiler | f8.8 | Boiler flow water temperature(°C)
| 27 | Tdhw | f8.8 | DHW temperature(°C)
| 28 | Toutside | f8.8 | Outside temperature(°C)
| 29 | Tret | f8.8 | Return water temperature(°C)
| 30 | Tstorage | f8.8 | Solar storage temperature(°C)
| 31 | Tcollector | f8.8 | Solar collector temperature(°C)
| 32 | TflowCH2 | f8.8 | Flow water temperature CH2 circuit(°C)
| 33 | Tdhw2 | f8.8 | Domestic hot water temperature 2 (°C)
| 34 | Texhaust | s16 | Boiler exhaust temperature(°C)
| 35 | TboilerHeatExchanger | f8.8 | Boiler heat exchanger temperature(°C)
| 36 | BoilerFanSpeedSetpointAndActual | u8/u8 | Boiler fan speed Setpoint and actual value
| 37 | FlameCurrent | f8.8 | Electrical current through burner flame[μA]
| 38 | TrCH2 | f8.8 | Room temperature for 2nd CH circuit(°C)
| 39 | RelativeHumidity | f8.8 | Actual relative humidity as a percentage
| 40 | TrOverride2 | f8.8 | Remote Override Room Setpoint 2
| 41 | TdhwSetUBTdhwSetLB | s8/s8 | DHW Setpoint upper & lower bounds for adjustment(°C)
| 42 | MaxTSetUBMaxTSetLB | s8/s8 | Max CH water Setpoint upper & lower bounds for adjustment(°C)
| 43 | TdhwSet | f8.8 | DHW Setpoint(°C) (Remote parameter 1)
| 44 | MaxTSet | f8.8 | Max CH water Setpoint(°C) (Remote parameters 2)
| 45 | StatusVentilationHeatRecovery | flag8/flag8 | Master and Slave Status flags ventilation / heat - recovery
| 46 | Vset | -/u8 | Relative ventilation position (0-100%). 0% is the minimum set ventilation and 100% is the maximum set ventilation. 
| 47 | ASFflagsOEMfaultCodeVentilationHeatRecovery | flag8/u8 | Application-specific fault flags and OEM fault code ventilation / heat-recovery 
| 48 | OEMDiagnosticCodeVentilationHeatRecovery | u16 | OEM-specific diagnostic/service code for ventilation / heat-recovery system 
| 49 | SConfigSMemberIDCodeVentilationHeatRecovery | flag8/u8 | Slave Configuration Flags / Slave MemberID Code ventilation / heat-recovery 
| 50 | OpenThermVersionVentilationHeatRecovery | f8.8 | The implemented version of the OpenTherm Protocol Specification in the ventilation / heat-recovery system. 
| 51 | VentilationHeatRecoveryVersion | u8/u8 | Ventilation / heat-recovery product version number and type 
| 52 | RelVentLevel | -/u8 | Relative ventilation (0-100%)
| 53 | RHexhaust | -/u8 | Relative humidity exhaust air (0-100%) 
| 54 | CO2exhaust | u16 | CO2 level exhaust air (0-2000 ppm) 
| 55 | Tsi | f8.8 | Supply inlet temperature (°C) 
| 56 | Tso | f8.8 | Supply outlet temperature (°C) 
| 57 | Tei | f8.8 | Exhaust inlet temperature (°C) 
| 58 | Teo | f8.8 | Exhaust outlet temperature (°C) 
| 59 | RPMexhaust | u16 | Relative humidity exhaust air (0-100%) 
| 60 | RPMsupply | u16 | Supply fan speed in rpm
| 61 | RBPflagsVentilationHeatRecovery | flag8/flag8 | Remote ventilation / heat-recovery parameter transfer-enable & read/write flags 
| 62 | NominalVentilationValue | u8/- | Nominal relative value for ventilation (0-100 %) 
| 63 | TSPventilationHeatRecovery | u8/u8 | Number of Transparent-Slave-Parameters supported by TSP’s ventilation / heat-recovery 
| 64 | TSPindexTSPvalueVentilationHeatRecovery | u8/u8 | Index number / Value of referred-to transparent TSP’s ventilation / heat-recovery parameter. 
| 65 | FHBsizeVentilationHeatRecovery | u8/u8 | Size of Fault-History-Buffer supported by ventilation / heat-recovery 
| 66 | FHBindexFHBvalueVentilationHeatRecovery | u8/u8 | Index number / Value of referred-to fault-history buffer entry ventilation / heat-recovery 
| 67 | Brand | u8/u8 | Index number of the character in the text string ASCII character referenced by the above index number 
| 68 | BrandVersion | u8/u8 | Index number of the character in the text string ASCII character referenced by the above index number 
| 69 | BrandSerialNumber | u8/u8 | Index number of the character in the text string ASCII character referenced by the above index number 
| 70 | CoolingOperationHours | u16 | Number of hours that the slave is in Cooling Mode. Reset by zero is optional for slave
| 71 | PowerCycles | u16 | Number of Power Cycles of a slave (wake-up after Reset), Reset by zero is optional for slave 
| 72 | RFsensorStatusInformation | special/special | For a specific RF sensor the RF strength and battery level is written 
| 73 | RemoteOverrideOperatingModeHeatingDHW | special/special | Operating Mode HC1, HC2/ Operating Mode DHW
| 74 | RemoteOverrideFunction | flag8/- | Function of manual and program changes in master and remote room Setpoint 
| 75 | StatusSolarStorage | flag8/flag8 | Master and Slave Status flags Solar Storage 
| 76 | ASFflagsOEMfaultCodeSolarStorage | flag8/u8 | Application-specific fault flags and OEM fault code Solar Storage 
| 77 | SConfigSMemberIDcodeSolarStorage | flag8/u8 | Slave Configuration Flags / Slave MemberID Code Solar Storage 
| 78 | SolarStorageVersion | u8/u8 | Solar Storage product version number and type
| 79 | TSPSolarStorage | u8/u8 | Number of Transparent - Slave - Parameters supported by TSP’s Solar Storage
| 80 | TSPindexTSPvalueSolarStorage | u8/u8 | Index number / Value of referred - to transparent TSP’s Solar Storage parameter.
| 81 | FHBsizeSolarStorage | u8/u8 | Size of Fault - History - Buffer supported by Solar Storage
| 82 | FHBindexFHBvalueSolarStorage | u8/u8 | Index number / Value of referred - to fault - history buffer entry Solar Storage
| 83 | ElectricityProducerStarts | u16 | Number of start of the electricity producer.
| 84 | ElectricityProducerHours | u16 | Number of hours the electricity produces is in operation
| 85 | ElectricityProduction | u16 | Current electricity production in Watt.
| 86 | CumulativElectricityProduction | u16 | Cumulative electricity production in KWh.
| 87 | UnsuccessfulBurnerStarts | u16 | Number of un - successful burner starts
| 88 | FlameSignalTooLowNumber | u16 | Number of times flame signal was too low
| 89 | OEMDiagnosticCode | u16 | OEM - specific diagnostic / service code
| 90 | SuccessfulBurnerStarts | u16 | Number of succesful starts burner
| 91 | CHPumpStarts | u16 | Number of starts CH pump
| 92 | DHWPumpValveStarts | u16 | Number of starts DHW pump / valve
| 93 | DHWBurnerStarts | u16 | Number of starts burner during DHW mode
| 94 | BurnerOperationHours | u16 | Number of hours that burner is in operation(i.e.flame on)
| 95 | CHPumpOperationHours | u16 | Number of hours that CH pump has been running
| 96 | DHWPumpValveOperationHours | u16 | Number of hours that DHW pump has been running or DHW valve has been opened
| 97 | DHWBurnerOperationHours | u16 | Number of hours that burner is in operation during DHW mode
| 98 | OpenThermVersionMaster | f8.8 | The implemented version of the OpenTherm Protocol Specification in the master.
| 99 | OpenThermVersionSlave | f8.8 | The implemented version of the OpenTherm Protocol Specification in the slave.
| 100 | MasterVersion | u8/u8 | Master product version number and type
| 101 | SlaveVersion | u8/u8 | Slave product version number and type

**Comment by grodansparadis** [comment]
As you wrote initially VSCP can do all of these as it stands today.  The thing with this protocol, and most other,  is the low abstraction level.  Nevertheless it works good for it's purpose and does the job. But  why use VSCP then and not stick with a working and mature protocol instead of a marginal unknown protocol like VSCP.  I may be wrong. I often is. Convince me if you thing thta is the case.

**Reply by michpro** [reply]
HVAC is just one subsystem of a larger whole. You've made the VSCP so that it has a broader perspective (and I like that about it, it's a shame I found it so hard to find, and it's not more popular). Including the HVAC class as a subset of the VSCP would allow for greater interoperability and unambiguous identification of events from such devices across the network. Now if, for example, I wanted to set the temperature of radiators or hot water I would most likely send a VSCP_TYPE_CONTROL_BIG_CHANGE_LEVEL event, which is very abstract, but I cannot indicate in it that I am setting the temperature, and its fractional values need to be described in the documentation, because I don't know if this sent number should be divided by 1, 10, 100,.... or maybe multiplied. I also do not know whether the unit is degrees Celsius, Fahrenheit or Kelvin. And so I would know all this without looking at the documentation. Using the measurement class (VSCP_TYPE_MEASUREMENT_TEMPERATURE) for this will in turn be confusing, because I am not sending the value I measured, I just want to force it to reach that state. Another case: currently if I want to quickly identify that a VSCP event came from some HVAC device I should put it in a common zone/subzone, and so the HVAC class would allow to clearly identify that the event came from e.g. a fridge in the kitchen zone, or a boiler in the zone - boiler room, and maybe - thermostat in the living room zone. But maybe that's just my weird craving. But I would still like to do it better with VSCP events, because it is a better, more flexible and cooler protocol than the others :) And I'm impatient with documentation :)

**Comment by grodansparadis** [comment]
First of all I have to express how glad I am for disussions like this.  It has not been a lot of that for some years now.  Much of VSCP comes from discussions like this and a lot of people wanting  to do things just like you are here.  

Generally I think that if it is obvious that a functionality is already there (in VSCP) it should not repetead. Or at least if there is not a good reason for doing so.  Measurements are a typical example of that. If creators try to use measurement classes when they send out measuremnet values it is simple en logic for others to understand and for other devices to hook on  in a plug and play way.  You can add a display, logger etc in a simple way and as long as the new device sending the data use VSCP it can be interchanged by another brand or type and everything will still work.

So the same is true for setting a value.  CLASS1.SETVALUEZONE can be used to set  values  in a typesafe maner.  Actually I got a little confused here, why is'nt there a CLASS1.SETVALUE? But it is't there beacuse of the lack of "addressing". Anyway CLASS1.SETVALUEZONE can be used to set a temperature for instance. It is a bit limited in value size and precission with two bytes less for "the value" compared to a measurement but will work for most situations.  I would definitly use it though to set the temperature of radiators or hot water in cases like yours.  For myself I usually end up setting setpoints in registers anyway. 

In your last case. The originating device is clearly identified by the GUID, or the nickname.  GUID and nickname is tightly coupled so once found/read from a device GUID will uniqully identify a device. So you know what device that was sending the measurement, this is used in decision matrices also.  For example I have a booiler controller in our house [Raspberry Pi 5](https://github.com/grodansparadis/vscp/wiki/VSCP-Demo-GUID%27s). A [Kelvin NTC10K board](https://www.grodansparadis.com/kelvinntc10k/kelvin_ntc10ka.html) measures temperature in the boiler and send out Turn on/turn off events continiously around the configured setpoint. The decision matrix of a  [Paris module](http://www.grodansparadis.com/paris/paris.html) triggers on the events and set/release the relay to the heater and confirm the state of the relay.  You can actually see this live with

```bash
mosquitto_sub -h mqtt.vscp.org -u vscp -P secret -t vscp/25:00:00:00:00:00:00:00:00:00:00:00:06:01:00:03/#
```

When I want to change the setpoint I have a web-application that writes the new register values. Same if I want to turn things on or off. I use node-red to display and log values for most of my stuff.  It would be easy to add a display that shows current values.

For manuals we all hate them.  And writing them is even harder because you know everyone hates them and still needs them and noone can find the things they are looking for in the anyway... :)

HVAC calls is great if we can add events to it that are not general for the VSCP protocol and if they contribute to the workings.  And again, if I don't get what you are speaking of, donät give up. I have a thick and may need some convincing... :)
